### PR TITLE
pageserver: flush deletion queue on detach

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

If a caller detaches a tenant and then attaches it again, pending deletions from the old attachment might not have happened yet.  This is not a correctness problem, but it causes:
- Risk of leaking some objects in S3
- Some warnings from the deletion queue when pending LSN updates and pending deletions don't pass validation.

## Summary of changes

- Deletion queue now uses UnboundedChannel so that the push interfaces don't have to be async.
  - This was pulled out of https://github.com/neondatabase/neon/pull/5397, where it is also useful to be able to drive the queue from non-async contexts.
  - Why is it okay for this to be unbounded?  The only way the unbounded-ness of the channel can become a problem is if writing out deletion lists can't keep up, but if the system were that overloaded then the code generating deletions (GC, compaction) would also be impacted.
- DeletionQueueClient gets a new `flush_advisory` function, which is like flush_execute, but doesn't wait for completion: this is appropriate for use in contexts where we would like to encourage the deletion queue to flush, but don't need to block on it.
  - This function is also expected to be useful in next steps for seamless migration, where the option to flush to S3 while transitioning into AttachedStale will also include flushing deletion queue, but we wouldn't want to block on  that flush.
- The tenant_detach code in  mgr.rs invokes flush_advisory after stopping the `Tenant` object.

### CODE REVIEW

**Review by arpad-m — COMMENTED** [review]
Needs a rebase.

Right now we would have requests get stuck and eventually time out if the queue gets stuck, which creates noisy errors.

I also wonder if we should somehow measure the queue length in a metric or warn if the queue length is larger than some limit. There is the risk of having a memory "leak" like the one fixed in #5472 ...

**Comment by jcsp** [comment]
> I also wonder if we should somehow measure the queue length in a metric or warn if the queue length is larger than some limit. There is the risk of having a memory "leak" like the one fixed in https://github.com/neondatabase/neon/pull/5472 ...

We do already have metrics that enable calculating queue length  -- this was a good reminder for me to add charts for those to the pageserver dashboard.  I'll follow up to add an alert for queue length.

**Comment by jcsp** [comment]
Push:
- Merge main into this branch
- Since the LocationConf stuff merged, we should also do this flush-on-detach behavior when we shut down a tenant in `upsert_location` -- added that.

File: pageserver/src/tenant/mgr.rs [hunk 1]
```diff
@@ -695,6 +694,18 @@ pub(crate) async fn upsert_location(
 
     if let Some(tenant) = shutdown_tenant {
         let (_guard, progress) = utils::completion::channel();
+
+        match tenant.get_attach_mode() {
+            AttachmentMode::Single | AttachmentMode::Multi => {
+                // Before we leave our state as the presumed holder of latest generation,
```
Comment (arpad-m) [inline pageserver/src/tenant/mgr.rs:700]:
```suggestion
                // Before we leave our state as the presumed holder of the latest generation,
```

**Review by arpad-m — APPROVED** [review]

**Pull request merged**